pi5/pinctrl_utils.go

```diff
@@ -0,0 +1,231 @@
+		}
+
+		rangeByteContents = rangeByteContents[(4 * numAddrSpaceCells):]
+		rangeInfo := rangeInfo{childAddr: childAddr, parentAddr: parentAddr, addrSpaceSize: addrSpaceSize}
+		addrRangesSlice = append(addrRangesSlice, rangeInfo)
+
+	}
+
+	return addrRangesSlice, err
+}
+
+// Uses Information Stored within the 'reg' property of the child node and 'ranges' property of its parents to map the child's physical address into the dev/gpiomem space
+func (b *pinctrlpi5) setGPIONodePhysAddr(nodePath string) error {
+
+	currNodePath := dtBaseNodePath + nodePath // initially: /proc/device-tree/axi/pcie@120000/rp1/gpio@d0000
+	var numCAddrCells uint32 = 0
+	var err error = nil
+
+	/* Call Recursive Function to Calculate Phys Addr. Works way up the Device Tree, using the information
+	found in #ranges at every node to translate from the child's address space to the parent's address space
+	until we get the child's physical address in all of /dev/gpiomem. */
+	*b.physAddr, err = setGPIONodePhysAddrHelper(currNodePath, INVALID_ADDR, numCAddrCells)
+	if err != nil {
+		return fmt.Errorf("trouble calculating phys addr for %s: %w\n", nodePath, err)
+	}
+
+	return nil
+}
+
+// Recursively Traverses Device Tree to Calcuate Physical Address of specified GPIO Chip
+func setGPIONodePhysAddrHelper(currNodePath string, physAddress uint64, numCAddrCells uint32) (uint64, error) {
+
+	if currNodePath == dtBaseNodePath { // Base Case: We are at the root of the device tree.
+		return physAddress, nil
+	}
+
+	// Normal Case: We are not at the root of the device tree. We must continue mapping our child addr (from the previous call) to this parent's addr space.
+	parentNodePath := filepath.Dir(currNodePath)
+	numPAddrCells, numAddrSpaceCells, err := getNumAddrSizeCellsInfo(parentNodePath)
+	if err != nil {
+		return INVALID_ADDR, err
+	}
+
+	var addrRangesSlice []rangeInfo
+	if physAddress == INVALID_ADDR { // Case 1: We are the Child Node. No addr has been set. Read the reg file to get the physical address within our parents space.
+		physAddress, err = getRegAddr(currNodePath, numPAddrCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+	} else { // Case 2: We use the ranges property to continue mapping a child addr into our parent addr space.
+		addrRangesSlice, err = getRangesAddrInfo(currNodePath, numCAddrCells, numPAddrCells, numAddrSpaceCells)
+		if err != nil {
+			return INVALID_ADDR, err
+		}
+
+		for _, addrRange := range addrRangesSlice {
+
+			if addrRange.childAddr <= physAddress && physAddress <= addrRange.childAddr+addrRange.addrSpaceSize {
+				physAddress -= addrRange.childAddr  // get the offset beween the address and child base address
+				physAddress += addrRange.parentAddr // now address has been mapped into parent space.
+				break
+			}
+		}
+	}
+
+	numCAddrCells = numPAddrCells
+	currNodePath = parentNodePath
+	return setGPIONodePhysAddrHelper(currNodePath, physAddress, numCAddrCells)
+}
```

Comment:
wont let me comment on the file but remove the pinctrl binary before merging this PR. Also remove the .local files on the main branch. you could add these to your gitignore so that they dont get pushed